### PR TITLE
James 1715

### PR DESCRIPTION
**Pull request by aduprat**
Only JAMES-1715 commits should be reviewed in this PR.
Other commit are from #200

### CODE REVIEW

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java [hunk 1]
```diff
@@ -59,4 +58,49 @@ public void fromShouldReturnSomethingWhenXPrefixedRole() {
         assertThat(Role.from("x-client-specific-role")).isEqualTo(Optional.of(new Role("x-client-specific-role")));
     }
 
+    @Test
+    public void isSystemRoleShouldTrueWhenInbox() {
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: s/ShouldTrue/ShouldBeTrue/
(and in next functions)

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java [hunk 2]
```diff
@@ -59,4 +58,49 @@ public void fromShouldReturnSomethingWhenXPrefixedRole() {
...
+        assertThat(Role.SPAM.isSystemRole()).isTrue();
+    }
+
+    @Test
+    public void isSystemRoleShouldTrueWhenTemplates() {
+        assertThat(Role.TEMPLATES.isSystemRole()).isTrue();
+    }
+
+    @Test
+    public void isSystemRoleShouldFalsWhenUserDefinedRole() {
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: s/ShouldFals/ShouldBeFalse/

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java [hunk 3]
```diff
@@ -501,4 +503,164 @@ public void setMailboxesShouldReturnNotCreatedWhenCycleDetected() throws Excepti
...
+        .then()
+            .log().ifValidationFails()
+            .statusCode(200)
+            .body(NAME, equalTo("mailboxesSet"))
+            .body(ARGUMENTS + ".notDestroyed", aMapWithSize(1))
+            .body(ARGUMENTS + ".notDestroyed", hasEntry(equalTo(mailboxId), Matchers.allOf(
+                    hasEntry(equalTo("type"), equalTo("invalidArguments")),
+                    hasEntry(equalTo("description"), equalTo("The mailbox '" + mailboxId + "' is a system mailbox.")))));
+    }
 }
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Can you add a test where you destroy a child and its parent?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/mailbox/RoleTest.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMailboxesMethodTest.java]: Ok

**Comment by aduprat** [comment]
test this please

**Comment by rouazana** [comment]
:+1:

**Comment by aduprat** [comment]
test this please

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    private void checkForChild(String id, MailboxSession mailboxSession) throws MailboxHasChildException, MailboxException {
+        if (mailboxUtils.hasChildren(id, mailboxSession)) {
+            throw new MailboxHasChildException();
+        }
+    }
+
+    private void checkRole(Optional<Role> role) throws SystemMailboxException {
+        if (role.isPresent() && role.get().isSystemRole()) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java]: more Optional friendly ?
```java
role.map(Role::isSystemRole).orElse(false)
```

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java]: Yes

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxUtils.java [hunk 5]
```diff
@@ -116,4 +116,36 @@ private String getMailboxId(MailboxPath mailboxPath, MailboxSession mailboxSessi
...
+        if (mailbox.getParentId().isPresent()) {
+            return getMailboxName(mailboxFromMailboxId(mailbox.getParentId().get(), mailboxSession).get(), mailboxSession) +
+                    mailboxSession.getPathDelimiter() + mailbox.getName();
+        }
+        return mailbox.getName();
+    }
+
+    public boolean hasChildren(String mailboxId, MailboxSession mailboxSession) throws MailboxException {
+        Optional<org.apache.james.mailbox.store.mail.model.Mailbox<Id>> mailbox = getMailboxFromId(mailboxId, mailboxSession);
+        if (mailbox.isPresent()) {
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxUtils.java]: could either be easily replaced with:
- getMailboxFromId(mailboxId, mailboxSession).map(Throwing...).orElse(false);
- or (at least) return mailbox.isPresent() && mailboxMapperFactory...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    private final CollectionHierarchySorter<Map.Entry<String, Mailbox>, String> collectionHierarchySorter;
+    private final MailboxUtils<Id> mailboxUtils;
+
+    @Inject
+    @VisibleForTesting
+    SetMailboxesDestructionProcessor(MailboxManager mailboxManager, MailboxUtils<Id> mailboxUtils) {
+        this.mailboxManager = mailboxManager;
+        this.collectionHierarchySorter =
+            new CollectionHierarchySorter<>(
+                    x -> x.getKey(),
```
Comment (fvn-linagora) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java]: favor method reference ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxUtils.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/utils/MailboxUtils.java]: Right

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesDestructionProcessor.java]: OK

**Comment by aduprat** [comment]
merged

**Pull request closed without merging**